MM-48582 Playbook import via drag and drop

m-ripper opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository-specific documentation at https://developers.mattermost.com/contribute/getting-started/

REMEMBER TO:
- Run `make i18n-extract` and commit changes to synchronize any new or removed messages
- Run `make check-style` to check for style errors (required for all pull requests)
- Run `make test` to ensure unit tests passed
-->

## Summary
<!--
A description of what this pull request does
-->
Playbooks can be imported by dropping a file onto the import buttons.

Additionally, the LHS is now refreshed when a playbook is imported.

[Screencast from 12.01.2023 16:54:38.webm](https://user-images.githubusercontent.com/8669935/212115666-a5733ea7-eb80-493c-815a-ba9f1cb82f18.webm)


## Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.:

  Fixes: https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the Jira ticket.
-->
[MM-48582](https://mattermost.atlassian.net/browse/MM-48582)
[MM-48591](https://mattermost.atlassian.net/browse/MM-48591)

## Checklist
<!-- Check off items as they are completed. ~~Strike through~~ items if they don't apply -->
- [ ] ~~Telemetry updated~~
- [ ] ~~Gated by experimental feature flag~~
- [x] Unit tests updated


[MM-48582]: https://mattermost.atlassian.net/browse/MM-48582?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

trilopin commented on tests-e2e/cypress/integration/playbooks/list_spec.js:
```diff
@@ -107,6 +107,67 @@ describe('playbooks > list', () => {
+            cy.findByTestId('titlePlaybook').within(() => {
+                // # Drop loaded fixture onto import button
+                cy.findByText('Import').selectFile('@playbookExport', {
+                    action: 'drag-drop'
```
TIL 🎉

trilopin commented on webapp/src/components/backstage/import_playbook.tsx:
```diff
@@ -8,37 +8,56 @@ import {importFile} from 'src/client';
+        if (!e.target.files?.[0]) {
+            return;
         }
+        readFile(e.target.files[0])
```
would make sense to limit the file size before reading the file and handle it as a user-readable error?

m-ripper replied:
That's probably a good idea. What file size were you thinking about if you had any in mind so far?

trilopin commented on webapp/src/components/backstage/import_playbook.tsx:
```diff
@@ -8,37 +8,56 @@ import {importFile} from 'src/client';
+    };
+
+    const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+        if (!e.target.files?.[0]) {
```
what happens if the user drop more than one file?

trilopin replied:
I tested, the first is imported, rest are ignored. Could we fail with an explicit error when more than 1 file is uploaded?

trilopin reviewed (changes requested):
Good job Max,

How hard would be to do the style change to visually tell the user that the import button is the drop area when drag start?

Please check the tiny details I wrote.

m-ripper commented:
I've added a visual indicator for the dropzone and updated the video above. I'm 2/5 in regards of the UX that the dropzone gets shown when the user drags a file to any part of the page. In Channels, it is only shown when dragged into the dropzone.

m-ripper commented on webapp/src/components/assets/files_overlay_image.tsx:
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
+
+const source = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAMAAAABrCAMAAAAW2X0YAAAC9FBMVEUAAAAAAAAAAAD////Ozs4AAAABAQHPz8/GxsYAAADMzMwAAADv7+/6+vri4uIAAAAAAAD7+/v////+/v78/Pz29vbNzc0AAAAAAAAAAAD9/f3+/v7+/v77+/v29vafn58AAAAAAABCQkL+/v7+/v7r6+vz8/MAAADJycn6+vr8/Pz9/f38/Pz39/f6+vr8/Pz8/Pzo6OjKysq4uLjd3d3////+/v739/f39/f19fXMzMzLy8umpqb////29vb39/f7+/vR0dHMzMzGxsYAAADd3d3JycmmpqYAAAD39/ft7e3Gxsa9v ...
```
I don't like this solution, but didn't find a better one at the moment. I want to use [this image](https://github.com/mattermost/mattermost-webapp/blob/master/images/filesOverlay.png) that is already included in the webapp but due to hashes I cannot use the already existing one. 

The Plugin API doesn't seem to support adding static files from a plugin, that's why I am using a base64 version of the the image, but maybe I am missing something. If there's a SVG version of the image, I will use it instead.

calebroseland replied:
I pushed a small fix to the webpack config that should support PNGs now—same outcome as what you have, just avoids having to do it manually. I think they have to be inlined because I think the current mattermost-server plugin build/flow that playbooks is in doesn't support URL resources (or other webpack chunks).

calebroseland commented on webapp/src/components/assets/files_overlay_image.tsx:
```diff
@@ -0,0 +1,14 @@
+
+import React from 'react';
+
+const source = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAMAAAABrCAMAAAAW2X0YAAAC9FBMVEUAAAAAAAAAAAD////Ozs4AAAABAQHPz8/GxsYAAADMzMwAAADv7+/6+vri4uIAAAAAAAD7+/v////+/v78/Pz29vbNzc0AAAAAAAAAAAD9/f3+/v7+/v77+/v29vafn58AAAAAAABCQkL+/v7+/v7r6+vz8/MAAADJycn6+vr8/Pz9/f38/Pz39/f6+vr8/Pz8/Pzo6OjKysq4uLjd3d3////+/v739/f39/f19fXMzMzLy8umpqb////29vb39/f7+/vR0dHMzMzGxsYAAADd3d3JycmmpqYAAAD39/ft7e3Gxsa9v ...
```
e.g. 
```suggestion
import source from './files_overlay.png';
```

calebroseland commented on webapp/src/components/backstage/import_playbook.tsx:
```diff
@@ -8,37 +8,76 @@ import {importFile} from 'src/client';
+            });
+            reject(new Error('File size limit exceeded'));
+            return;
+        }
```
```suggestion
        }
        if (file.type !== 'application/json') {
            addToast({
                content: formatMessage({defaultMessage: 'The file must be a valid JSON playbook template.'}),
                toastStyle: ToastStyle.Failure,
            });
            reject(new Error('File must be a JSON file'));
            return;
        }
```

calebroseland reviewed (approved):
Awesome work @m-ripper! Thanks for the expanded indicators.

One thought on a helpful error message if you try to upload an unsupported file type, but not essential so feel free to proceed without.

abhijit-singh reviewed (approved):
Love it @m-ripper!

trilopin reviewed (approved):
Good job @m-ripper 

Non blocking suggestion regarding messages accuracy
